src/VisualStudio/Core/Def/Implementation/Options/RoamingVisualStudioProfileOptionPersister.cs

```diff
@@ -105,24 +124,30 @@ private object GetFirstOrDefaultValue(OptionKey optionKey, IEnumerable<RoamingPr
             // location (and can upgrade the value accordingly) but we only write to the new location so that
             // we don't interfere with older versions. This will essentially "fork" the user's options at the time of upgrade.
 
-            foreach (var roamingSerialization in roamingSerializations)
-            {
-                var storageKey = roamingSerialization.GetKeyNameForLanguage(optionKey.Language);
+            return ThreadingContext.JoinableTaskFactory.Run<object>(
```

Comment:
JTF probably isn't safe here

Reply:
Why not?

Reply:
I'm going to have to ask @jasonmalinowski / @sharwell (or anyone else in @dotnet/roslyn-ide with experience here) for help with the threading work.

Reply:
Alright, so this is a bit complicated.

The JTF.Run here is actually _fine_, in that it's not the problem to be worried about. This code is either being called on a JTF main thread, where this correctly is implementing JTF rules, or it's being called off the UI thread where deadlock concerns are more or less moot. But that's not the problem...

Previously to the best of my knowledge, ISettingsManager.TryGetValue was explicitly guaranteed to be free-threaded. This was a requirement we pushed for, because Roslyn has contractual guarantees that our public APIs that return Tasks do _not_ implicitly jump back to the UI thread, so you are safe to .Wait() them. We have problems here not because of JTF.Run() being used here, but rather if GetValueAsync() here is implicitly requiring the UI thread at any point during it's operation. No JTF.Run() is going to save us there if the original caller is using a JTF-unfriendly wait, and this will deadlock regardless. Combining this all, it somewhat means that either:

1. The JTF.Run() isn't required, because the IBrokeredSettingsManager.GetValueAsync ensures no hops back to the UI thread as a part of it's operation. Put another way, it makes the same guaranteed as before. Therefore we have no concerns about violating the existing Roslyn deadlock guarantees, and we'll pinch our nose at the blocking wait here.
2. The JTF.Run() is required because IBrokeredSettingsManager.GetValueAsync() may indeed hop back to the UI thread. It's entitled to of course if this is entirely in-proc, but then it means we're potentially violating the Roslyn deadlock guarantee and so this code is already toast, no matter what we do in this method.

(Now, if fetching an option requires a hop to the UI thread I have *other* performance concerns as well, because we very well may read a huge number of options at the very start and that's just not good for perf!)

So I guess a question to @JoshuaBStevens: which is it? "It could be either" is of course a valid answer too, which just means we have the pain of both!

Now there's a much bigger concern really: we're blocking trying to fetch an option. That seems bad. If this is being called *on the client* and it's directly calling the server to do this, this is going to be a source of massive perf issues, so it seems like we need to do something else there. So I presume this is only being called on the server, where maybe we know this is always fast? If it's always fast we can maybe get away with this, but if it's not fast there that's still a huge issue.

So I have a second question for @JoshuaBStevens: I open Tools > Options, and click on one of the Roslyn pages. Has the brokered service already mirrored all settings down to the client, or are we going to be fetching those values when I click the page? How are we dealing with the UI lag at this point?

Reply:
(I wonder too more generally: what scenarios other than option pages does the client in Roslyn need to read a server option? Are all the reads actually happening on the server as a part of a request?)

Reply:
The policy is "no blocking on brokered service calls or service requests". The rule doesn't specify whether it's main thread or not because when it comes to how long a network call can last, even a threadpool thread is a precious resource.
Can we avoid blocking entirely?

Reply:
IBrokeredSettingsManager.GetValueAsync does not require the UI thread.  As to your second question, we do NOT mirror settings from the Nexus server down to the client, so getting them might incur a UI delay.  If this ends up being an issue, IMO the proper way to address it would be to show some kind of non-blocking progress UI in place of the Options page until the settings are downloaded.  If necessary I will implement support for that in the DialogPage base class.

Reply:
> If this ends up being an issue, IMO the proper way to address it would be to show some kind of non-blocking progress UI in place of the Options page until the settings are downloaded. If necessary I will implement support for that in the DialogPage base class.

What are the other pages that fetch from the server doing today if there isn't already support in DialogPage?

Reply:
DialogPage has support for loading settings from the Nexus server when appropriate; what's missing is support for presenting a non-blocking progress UI while the load is occurring.

Reply:
So what is today's experience? It just hangs the UI thread during the load?

Reply:
Yes.

Reply:
> IBrokeredSettingsManager.GetValueAsync does not require the UI thread

Does any part of `IBrokeredSettingsManager` workflow require UI thread? If not then let's remove `ForegroundThreadAffinitizedObject` and all JTF usage from this class.

Reply:
GetSettingInfo[s] may switch to the UI thread to lazily load localized resources.  SetValueAsync will call event handlers which may require the UI thread.  Other than that, no methods on IBrokeredSettingsManager have UI thread dependencies.

Reply:
Would it be possible to change these APIs (especially SetValueAsync) to not dispatch to UI thread? It doesn't make much sense to me that a key-value store requires UI thread to read/write data.

Reply:
The store doesn't require the UI thread, but subscribers to the setting change event *may* require the UI thread, and SetValueAsync executes those event handlers before returning.  (Async event handlers are supported via ISettingsManager.GetSubset(…).SettingChangedAsync.)  I suspect you would like a way to wait for the setting to be persisted but NOT wait for change event handlers to execute; is that accurate?

Reply:
Exactly. Essentially have a code path that does not dispatch to UI thread and can be completely free-threaded.

Reply:
I opened DevDiv bug 1051756 to track this request.

Reply:
I would have imagined SetValueAsync just always queued the notifications for events asynchronously and didn't block -- do you have scenarios where people need the guarantee? Discussing it now I can imagine some patterns that might expect it; I just wouldn't have thought of it at first.

Reply:
IIRC there were some internal scenarios (i.e. code I own) where I needed this.  I don't think any other teams have requested it.  I think, especially in the Nexus world, it makes sense not to block by default.  I'm inclined to change the behavior of IBrokeredSettingsManager to not block (possibly with an optional override to force blocking) and leave the default behavior of ISettingsManager as blocking (again, possibly introducing an optional override) for compatibility reasons.

Reply:
I prefer the pattern where event handlers never block the callers who made the change, but that if folks want that, you can return a 'version number' from the Set method, and another method to `WaitForEventsTillAsync(int version)`. This is what CPS does, and it's *very* flexible.